### PR TITLE
Move support for sub-flows to actions

### PR DESCRIPTION
**Pull request by jorgenj**
Move support for sub-flows to actions, so we get all the other features of actions for sub-flows (like looping, retries, error-handling, etc)

**Many thanks for submitting your Pull Request :heart:!**

**Please specify parts this PR updates:**

- [✅ ] Specification
- [✅ ] Schema
- [✅ ] Examples
- [ ] Extensions
- [ ] Roadmap
- [ ] Use Cases
- [ ] Community
- [ ] TCK
- [ ] Other

**What this PR does / why we need it**:

The intent is to unify the experience for using actions & sub-flows, since sub-flow use-cases have many similar needs as actions.

https://github.com/serverlessworkflow/specification/issues/300

> To give an example of what I mean, I think we could do the following and the language would be much more consistent:
>
> Add 'subFlow' as an optional parameter for an Action definition. This would be an object with 2 parameters, one is the sub-workflow ID and the other a waitForCompletion flag.
Drop the SubFlow state from the spec
Drop the 'workflowId' parameter from parallel-state branch definition
With that change, a customer can use sub-flows like they can invoke any other type of action in parallel, operation, & other states. Existing retry behavior could now be reused for sub-workflows, looping constructs that are available for actions are also automatically supported.
>
> Really, I see very little difference in behavior between a sub-workflow & the eventRef/functionRef style of action (under the covers all invoke something and wait for a result back). In the case that a user sets waitForCompletion=false, then the spec should indicate that the parent workflow should behave as if the sub-workflow completed immediately with an output value of an empty json object perhaps.
>
> A good illustration of my point is that if you compare the current description of Operation and SubFlow states, there seems to be very little difference in their features/behavior.

**Special notes for reviewers**:

**Additional information (if needed):**

### CODE REVIEW

File: comparisons/comparison-bpmn.md
Comment (tsurdilo) [inline comparisons/comparison-bpmn.md:316]: ForEach state is incorrect here imo. 
Foreach state executes each iteration in parallel - https://github.com/serverlessworkflow/specification/blob/main/specification.md#foreach-state --> "Each iteration of the ForEach state should be executed in parallel."

where "repeat" is a loop one at a time

also in jq you can use "[range(0;9)]" as the expression instead of creating the array manually 

So I think foreach state is not a replacement for repeat

File: examples/README.md
Comment (tsurdilo) [inline examples/README.md:3349]: if we allow async exec for subflow actions it should be also possible for functionRef actions

File: examples/README.md
Comment (tsurdilo) [inline examples/README.md:3359]: Don't understand the logic here. The intend is to repeat execution of the subflow until the car stopped event is consumed. this seems its just doing it once and transitioning to this event state. am i missing something?

Comment (tsurdilo) [reply examples/README.md:3359]: ah ok there are more states below.

**Comment by tsurdilo** [comment]
I think moving subflow invocation in actions is a good idea as discussed in meeting.

Don't think removing repeat without an adequate alternative tho is a good thing (as we had the community request for it).
We can take a look at allowing repeating (loop) in actions as foreach state is not an alternative for it.

Do any of the example images need updates for this ?

File: comparisons/comparison-bpmn.md
Comment (tsurdilo) [reply comparisons/comparison-bpmn.md:316]: also i think this implies that input to the inputCollection is an empty object where as the input to the inputCollection is the state data 
(meaning the expression in the pr or using "range" only work on "{ }", they don't just generate an array out of the blue)

Comment (tsurdilo) [reply comparisons/comparison-bpmn.md:316]: just noticed the bpmn image is incorrect, in that it shows the multi-instance subprocess instead of the loop marker. That probably caused the confusion, sorry. Will update the image asap

**Comment by ricardozanini** [comment]
@tsurdilo @jorgenj, can we leave the `repeat` topic off this PR? So we can discuss it further?

**Comment by tsurdilo** [comment]
@ricardozanini I am not sure if that's possible as this pr removes the subflow state and thus removes the repeat/loop functionality as well. i think we can definitely use this pr with the addition of repeat to actions. Just have to figure out what the best way is to do that

**Comment by tsurdilo** [comment]
updated bpmn2 examples - https://github.com/serverlessworkflow/specification/blob/main/comparisons/comparison-bpmn.md (sorry this one will need a small rebase)

File: comparisons/comparison-bpmn.md
Comment (jorgenj) [reply comparisons/comparison-bpmn.md:316]: You ignored that I put `max: 1`.  This makes it serial in nature...

Comment (tsurdilo) [reply comparisons/comparison-bpmn.md:316]: ok i see, missed it. i guess then the issue left is  that data input for each subflow exec here is the element of the input collection (1, 2, 3, 4 ...) and not state data as intended by business logic in the initial example.

Comment (tsurdilo) [reply comparisons/comparison-bpmn.md:316]: i think i removed a comment here too somehow :( 
inputCollection expression data input is the state data, so i think it cannot be an expression that somehow creates the data input for the subflow executions. either way the expressions have to operate on some data input and i assume the intend was that they somehow operate on "{ }" maybe, idk. for example it guess just like ${ .orders } or even ${ .inputArray } should be fine for users to undertand

Comment (jorgenj) [reply comparisons/comparison-bpmn.md:316]: @tsurdilo I'm not sure I follow what you're saying in your last comment.

I can adjust the expression so that it's iterating 10 times and the input to each iteration is the state input with the following, I believe:

```yaml
inputCollection: ${ reduce range(0,10) as $d ([.]; .[$d+1] += .[0] ) }
```

That's probably not very easy to understand at first glance though.   Are you suggesting that I just change the example to say there's an array of 'orders' and the example just iterates over those 'orders'?

Comment (jorgenj) [reply comparisons/comparison-bpmn.md:316]: Got more clarity, will replace 'repeat' usage in examples with explicit control-flow logic.

**Comment by tsurdilo** [comment]
@jorgenj there are some conflicts that should be resolved before merging. If you want I can fix them, just let me know It's np (I am a rebase pro ;) )

File: comparisons/comparison-bpmn.md
Comment (tsurdilo) [inline comparisons/comparison-bpmn.md:317]: I think here you dont have to use the inject state. Just assume that "count" (I would call it "max") is passed as workflow data input.

File: comparisons/comparison-bpmn.md
Comment (tsurdilo) [inline comparisons/comparison-bpmn.md:330]: I think it would be clearer if we move this to the above operation state (in stateDataFilter output param)

File: examples/README.md
Comment (tsurdilo) [inline examples/README.md:3519]: Received?

File: examples/README.md
Comment (tsurdilo) [inline examples/README.md:3518]: What is the ${ true } expression supposed to do?

File: examples/README.md
Comment (tsurdilo) [inline examples/README.md:3531]: Received

File: specification.md
Comment (tsurdilo) [inline specification.md:2272]: Should we not call this "subflowRef" instead so it fits nicely with "eventRef and "functionRef"?

File: specification.md
Comment (tsurdilo) [inline specification.md:4131]: inside a ForEach or Parallel state
(missing "state")

**Comment by tsurdilo** [comment]
@jorgenj added a couple of small comments. 
The only thing that I think is left is to made a decision on what to do with waitForCompletion in the subflow definition.
As we currently don't have a similar construct for functionRef it's weird to just have it on subflow.

We could address it with this PR or if you want with a separate one, up to you, let me know what you want to do.

**Comment by tsurdilo** [comment]
@jorgenj please take a look at https://github.com/serverlessworkflow/specification/issues/329
i think this is a very valid request. Let me know if you would like to add it to be part of this pr or a separate.

**Comment by jorgenj** [comment]
> @jorgenj please take a look at #329
> i think this is a very valid request. Let me know if you would like to add it to be part of this pr or a separate.

I'd like to get this merged and we can tackle that in a separate PR

**Comment by jorgenj** [comment]
> @jorgenj added a couple of small comments.
> The only thing that I think is left is to made a decision on what to do with waitForCompletion in the subflow definition.
> As we currently don't have a similar construct for functionRef it's weird to just have it on subflow.
> 
> We could address it with this PR or if you want with a separate one, up to you, let me know what you want to do.

I'd prefer if we tackle that in a separate PR, it's not clear to me where that property should reside.  On both functionRef and subflowRef or in some shared location?

File: comparisons/comparison-bpmn.md
Comment (jorgenj) [reply comparisons/comparison-bpmn.md:317]: Done

File: comparisons/comparison-bpmn.md
Comment (jorgenj) [reply comparisons/comparison-bpmn.md:330]: Done

File: examples/README.md
Comment (jorgenj) [reply examples/README.md:3518]: It's supposed to replace the results of the event, so that the value assigned to .stopReceived is 'true'.

File: specification.md
Comment (jorgenj) [reply specification.md:2272]: For functions and events, we have other objects in the same document that the functionRef/eventRef are referring to.  For subflow, it's a completely different doc so it seemed odd to me to call it subFlowRef.  It's a minor thing, I'll defer to your judgement and rename it subFlowRef.

File: examples/README.md
Comment (jorgenj) [reply examples/README.md:3518]: Switched to using has() in the switch state's data-condition.

**Comment by tsurdilo** [comment]
@jorgenj thank you for making the updates and rebasing! I wanted to ask one more thing please
I don't understand how the subflow "waitforcompletion" interacts with the actionMode property on actions which can be "sequential" or "parallel" currently. 
The issue is that when we set actionMode to "sequential" (which is the default) we guarantee that the results of the action will be merged to state data before the next action is executed.
With that and waitForCompletion set to false, I am not sure this fits that model any longer and imo can cause weird behaviour.

Let me know what you think. Thanks.

**Comment by jorgenj** [comment]
> @jorgenj thank you for making the updates and rebasing! I wanted to ask one more thing please
> I don't understand how the subflow "waitforcompletion" interacts with the actionMode property on actions which can be "sequential" or "parallel" currently.
> The issue is that when we set actionMode to "sequential" (which is the default) we guarantee that the results of the action will be merged to state data before the next action is executed.
> With that and waitForCompletion set to false, I am not sure this fits that model any longer and imo can cause weird behaviour.
> 
> Let me know what you think.
> Imo we should remove the waitForCompletion on subflow action for now and it would be nice if it was done alongside this PR please

@tsurdilo 

I'm fine removing it for now in this PR.  

Though, my interpretation is that when waitForCompletion is false, then the workflow should behave as if the sub-workflow (or action) completely immediately with an empty json object as the output of the subflow/action.   In that regard, then as far as actionMode=sequential is concerned it's really no different from a subflow/action that we waited for and which returned an empty json object as it's output.   WDYT?

**Comment by tsurdilo** [comment]
@jorgenj I think what you mentioned is fair, and we should probably make sure it is in the spec doc as well.
Had a chance to think about this for a bit and there are couple of things that are concerning me, some we already mentioned before like:
1) adding consistent loop/repeat ability for actions (which we said we would address in near future)
2) adding consistency to sync/async capabilities of all action types (functionRef, eventRef, and subflowRef). For this if we want to tackle this in a separate PR thats fine but I do think that we should remove waitForCompletion for now and address it across the board. One more thing that we should probably address here is the functionRef to a function definition of type expression. I think that either we should state in spec that referencing expression type functions should not be allowed in actions or if it is allowed to define that async invocation of those if probably not allowed.

3) (this one is new :) ) - if you look at the actions now, both functionRef and eventRef reference function and event definitions which are defined outside of the main "states" control flow logic definition, and allow users to define those as separate json/yaml which can be imported into the main workflow definition.
I believe that this should also be the case for subFlow actions, meaning that I would like to also have a "subflows" top-level definition which just like the others can be an in-line array or an URI to a separate file where those are defined.
This imo would help with consistency across different actions as well. WDYT?

**Comment by jorgenj** [comment]
> @jorgenj I think what you mentioned is fair, and we should probably make sure it is in the spec doc as well.
> Had a chance to think about this for a bit and there are couple of things that are concerning me, some we already mentioned before like:
> 
> 1. adding consistent loop/repeat ability for actions (which we said we would address in near future)

I'm not sure I understand what you'd like to be done in this PR.  Can you expand on this?

> 2. adding consistency to sync/async capabilities of all action types (functionRef, eventRef, and subflowRef). For this if we want to tackle this in a separate PR thats fine but I do think that we should remove waitForCompletion for now and address it across the board. 

Got it, I'll remove it for now then.

> One more thing that we should probably address here is the functionRef to a function definition of type expression. I think that either we should state in spec that referencing expression type functions should not be allowed in actions or if it is allowed to define that async invocation of those if probably not allowed.

I'm not sure I understand the connection to this PR.  This seems more like a future concern for the PR where async support for functions/subFlows is added?

> 3. (this one is new :) ) - if you look at the actions now, both functionRef and eventRef reference function and event definitions which are defined outside of the main "states" control flow logic definition, and allow users to define those as separate json/yaml which can be imported into the main workflow definition.
>    I believe that this should also be the case for subFlow actions, meaning that I would like to also have a "subflows" top-level definition which just like the others can be an in-line array or an URI to a separate file where those are defined.
>    This imo would help with consistency across different actions as well. WDYT?

That's an interesting idea, but I'm not sure I understand the benefit.  The subFlowRef is already a pretty light-weight 'object' referencing an external resource, I'm not sure how much benefit there would be in sharing those externally.

**Comment by jorgenj** [comment]
@tsurdilo

> > 1. adding consistency to sync/async capabilities of all action types (functionRef, eventRef, and subflowRef). For this if we want to tackle this in a separate PR thats fine but I do think that we should remove waitForCompletion for now and address it across the board.
> 
> Got it, I'll remove it for now then.

On second thought, there's a couple examples that would be invalidated and/or hard to reproduce if I remove `waitForCompletion`.  Is this a strong requirement that we *must* remove waitForCompletion for now?  It seems like busy-work if we're just going to be adding it back again later.

File: specification.md
Comment (jorgenj) [inline specification.md:2527]: From the meeting discussion, this section here talks about asynch sub-flow & how the parent workflow should behave in that case.

**Review by tsurdilo — APPROVED** [review]

**Pull request merged**
